File: dotnet/Source/eg/Page.java
// Copyright (c) 2002 Cunningham & Cunningham, Inc.
// Released under the terms of the GNU General Public License version 2 or later.

package eg;

import fit.*;
import java.net.*;
import java.io.*;
import java.util.StringTokenizer;

public class Page extends RowFixture {
    static URL url;
    static String text;

    // actions //////////////////////////////////

    public void location(String url) throws Exception {
        this.url = new URL(url);
        this.text = get(this.url);
    }

    public String title() throws Exception {
        String tags[] = {"title"};
        return new Parse(text, tags).text();
    }

    public void link (String label) throws Exception {
        String tags[] = {"a"};
        Parse links = new Parse (text, tags);
        while (!links.text().startsWith(label)) {
            links = links.more;
        }
        StringTokenizer tokens = new StringTokenizer(links.tag, "<> =\"");
        while (!tokens.nextToken().toLowerCase().equals("href")) {};
        url = new URL(url, tokens.nextToken());
        text = get(url);
    }

    // rows /////////////////////////////////////

    public Class getTargetClass() {
        return Row.class;
    }

    public Object[] query() throws Exception {
        String tags[] = {"wiki", "table", "tr", "td"};
        Parse rows = new Parse(text, tags).at(0, 0, 2);
        Row result[] = new Row[rows.size()];
        for (int i=0; i<result.length; i++) {
            result[i] = new Row(rows);
            rows = rows.more;
        }
        return result;
    }

    public class Row {
        Parse cells;

        Row(Parse row) {
            this.cells = row.parts;
        }

        public String numerator() {
            return cells.at(0).text();
        }

        public String denominator() {
            return cells.at(1).text();
        }

        public String quotient() {
            return cells.at(2).text();
        }
    }


    // utilities ////////////////////////////////

    private String get(URL url) throws IOException {
        this.url = url;
        InputStream stream =  (InputStream) url.getContent();
        BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
        StringBuffer buffer = new StringBuffer(10000);
        for (String line; (line = reader.readLine()) != null; ) {
            buffer.append(line);
            buffer.append('\n');
        }
        return buffer.toString();
    }



}

File: imp/cpp/src/eg/Page.cpp
/**
 * <p>This file is part of CeeFIT.</p>
 *
 * <p>CeeFIT is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.</p>
 *
 * <p>CeeFIT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.</p>
 *
 * <p>You should have received a copy of the GNU General Public License
 * along with CeeFIT; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA</p>
 *
 * <p>(c)2005 Woldrich, Inc.  <a href="http://www.woldrich.com">http://www.woldrich.com</a></p>
 *
 * @author David Woldrich
 */

#include "tools/alloc.h"
#include "ceefit.h"
#include "eg/eg.h"

using namespace CEEFIT;

namespace EG
{
  STRING PAGE::Url;
  STRING PAGE::Text;

  void ceefit_call_spec PAGE::Location(const STRING& url) 
  {
    PAGE::Url = url;
    PAGE::Text = Get(PAGE::Url);
  }

  STRING ceefit_call_spec PAGE::Title(void) 
  {
    DYNARRAY<CEEFIT::STRING> tags;
    tags.Add("title");
    PTR<PARSE> aParse(new PARSE(PAGE::Text, tags));
  
    return(aParse->Text());
  }

  void ceefit_call_spec PAGE::Link(const STRING& label) 
  {
    DYNARRAY<STRING> tags;
    tags.Add("a");
    PTR<PARSE> links(new PARSE(PAGE::Text, tags));

    while(!links->Text().StartsWith(label)) 
    {
      links = links->More;
    }
    DYNARRAY<STRING> tokens;
    Tokenize(tokens, links->Tag, STRING("<> =\""));
    int i = 0;
    while(i < tokens.GetSize() && !tokens[i].ToLowercase().IsEqual("href")) 
    {
      i++;
    }
    Url = Url + tokens[i+1];
    Text = Get(Url);
  }

  // rows /////////////////////////////////////

  void ceefit_call_spec PAGE::Query(DYNARRAY< PTR<CELLADAPTER> >& out) 
  {
    throw new EXCEPTION("Unimplemented function");
    /*
    String tags[] = {"wiki", "table", "tr", "td"};
    Parse rows = new Parse(text, tags).at(0, 0, 2);
    Row result[] = new Row[rows.size()];
    for (int i=0; i<result.length; i++) 
    {
      result[i] = new Row(rows);
      rows = rows.more;
    }
    return result;*/
  }

  /*
  class Row 
  {
    Parse cells;

    Row(Parse row) 
    {
      this.cells = row.parts;
    }

    STRING Numerator() 
    {
      return Cells->At(0)->Text();
    }

    STRING Denominator() 
    {
      return Cells->At(1)->Text();
    }

    public String quotient() {
      return cells.at(2).text();
    }
  }
  */

  // utilities ////////////////////////////////

  STRING ceefit_call_spec PAGE::Get(const STRING& url) 
  {
    throw new EXCEPTION("Unimplemented function");
    /*
    Page.url = url;
    InputStream stream =  (InputStream) url.getContent();
    BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
    StringBuffer buffer = new StringBuffer(10000);
    for (String line; (line = reader.readLine()) != null; ) {
      buffer.append(line);
      buffer.append('\n');
    }
    return buffer.toString();
    */
  }
};